pkg/controller/aws/cache/replicationgroup.go

```diff
@@ -333,31 +325,47 @@ func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error)
 	return reconcile.Result{Requeue: client.Sync(ctx, rd)}, errors.Wrapf(r.kube.Update(ctx, rd), "cannot update resource %s", req.NamespacedName)
 }
 
-func (r *Reconciler) upsertSecret(ctx context.Context, s *corev1.Secret) error {
-	n := types.NamespacedName{Namespace: s.GetNamespace(), Name: s.GetName()}
-	if err := r.kube.Get(ctx, n, &corev1.Secret{}); err != nil {
+func (r *Reconciler) upsertSecret(ctx context.Context, new *corev1.Secret) error {
+	n := types.NamespacedName{Namespace: new.GetNamespace(), Name: new.GetName()}
+	existing := &corev1.Secret{}
+	if err := r.kube.Get(ctx, n, existing); err != nil {
 		if kerrors.IsNotFound(err) {
-			return errors.Wrapf(r.kube.Create(ctx, s), "cannot create secret %s", n)
+			return errors.Wrapf(r.kube.Create(ctx, new), "cannot create secret %s", n)
 		}
 		return errors.Wrapf(err, "cannot get secret %s", n)
 	}
-	return errors.Wrapf(r.kube.Update(ctx, s), "cannot update secret %s", n)
+
+	// If a key is set in both the existing and new secrets the new key wins,
+	// but we preserve any keys set in the existing secret that aren't set in
+	// the new secret.
+	for k, v := range existing.Data {
+		if _, ok := new.Data[k]; !ok {
+			new.Data[k] = v
+		}
+	}
```

Comment:
There's an edge case here:

1. Create replication group `foo` with connection secret `foo` and authentication enabled.
1. Delete replication group `foo`, but don't delete the secret.
1. Create replication group `bar` with connection secret `foo` and authentication disabled.

We'll preserve the password from the existing connection secret, making it appear as if `bar` requires a password, when it does not. I'm not sure how concerning this case is in practice.